### PR TITLE
TEL-4736: add gateway register custom variables support

### PR DESCRIPTION
**Pull request by StefanYohansson**
Screenshot from 2021-03-02 15-43-51

### CODE REVIEW

File: src/mod/endpoints/mod_sofia/sofia_reg.c [hunk 1]
```diff
@@ -453,17 +454,25 @@ void sofia_reg_check_gateway(sofia_profile_t *profile, time_t now)
...
 							 SIPTAG_EXPIRES_STR(gateway_ptr->expires_str),
 							 NUTAG_REGISTRAR(gateway_ptr->register_proxy),
 							 NUTAG_OUTBOUND("no-options-keepalive"), NUTAG_OUTBOUND("no-validate"), NUTAG_KEEPALIVE(0), TAG_NULL());
 				gateway_ptr->retry = now + gateway_ptr->retry_seconds;
+
+				switch_safe_free(custom_headers);
+				custom_headers = NULL;
 			} else {
 				gateway_ptr->status = SOFIA_GATEWAY_DOWN;
 				nua_unregister(gateway_ptr->nh,
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_reg.c:478]: I think its better to support this feature for unregister?

File: src/mod/endpoints/mod_sofia/sofia_reg.c [hunk 2]
```diff
@@ -453,17 +454,25 @@ void sofia_reg_check_gateway(sofia_profile_t *profile, time_t now)
...
 							 SIPTAG_TO_STR(gateway_ptr->distinct_to ? gateway_ptr->register_to : gateway_ptr->register_from),
 							 SIPTAG_CONTACT_STR(gateway_ptr->register_contact),
 							 SIPTAG_FROM_STR(gateway_ptr->register_from),
 							 SIPTAG_EXPIRES_STR(gateway_ptr->expires_str),
 							 NUTAG_REGISTRAR(gateway_ptr->register_proxy),
 							 NUTAG_OUTBOUND("no-options-keepalive"), NUTAG_OUTBOUND("no-validate"), NUTAG_KEEPALIVE(0), TAG_NULL());
 				gateway_ptr->retry = now + gateway_ptr->retry_seconds;
+
+				switch_safe_free(custom_headers);
+				custom_headers = NULL;
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_reg.c:475]: Just minor but switch_safe_free is a macro that check nullity, free it then set the variable to null. No need to set the custom_headers to NULL here

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/mod/endpoints/mod_sofia/sofia_reg.c (hunk 1 above)
Comment (StefanYohansson) [reply src/mod/endpoints/mod_sofia/sofia_reg.c:479]: added 
```
#define SOFIA_SIP_UNREGISTER_HEADER_PREFIX "sip_unreg_h_"
``` 

thank you for the suggestion

**Review by pchero — APPROVED** [review]
Looks good to me.

File: src/mod/endpoints/mod_sofia/sofia_reg.c [hunk 3]
```diff
@@ -465,11 +471,16 @@ void sofia_reg_check_gateway(sofia_profile_t *profile, time_t now)
 							 NUTAG_OUTBOUND("no-options-keepalive"), NUTAG_OUTBOUND("no-validate"), NUTAG_KEEPALIVE(0), TAG_NULL());
 				gateway_ptr->retry = now + gateway_ptr->retry_seconds;
 			} else {
+				if (gateway_ptr->ib_vars) {
+					custom_headers = sofia_glue_get_extra_headers_from_event(gateway_ptr->ib_vars, SOFIA_SIP_REGISTER_HEADER_PREFIX);
+				}
+
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_reg.c:477]: Redundant code, Please move this on line 444 or in 455

File: src/mod/endpoints/mod_sofia/sofia_reg.c [hunk 4]
```diff
@@ -453,10 +454,15 @@ void sofia_reg_check_gateway(sofia_profile_t *profile, time_t now)
 			switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_NOTICE, "Registering %s\n", gateway_ptr->name);
 
 			if (now) {
+				if (gateway_ptr->ib_vars) {
+					custom_headers = sofia_glue_get_extra_headers_from_event(gateway_ptr->ib_vars, SOFIA_SIP_REGISTER_HEADER_PREFIX);
+				}
+
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_reg.c:460]: Redundant code, Please move this on line 444 or in 455

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

**Review by dev-ryanc — APPROVED** [review]

**Pull request merged**